src/tools/idlc/idlc.c

```diff
@@ -49,59 +49,55 @@ typedef struct {
 /* mcpp does not accept userdata */
 static int32_t retcode = 0;
 static idlc_options_t opts;
-static idl_processor_t proc;
+static idl_pstate_t *pstate = NULL;
 
 static int idlc_putc(int chr, OUTDEST od);
-static int idlc_puts(const char *str, OUTDEST od)
-  ddsrt_nonnull((1));
-static int idlc_printf(OUTDEST od, const char *str, ...)
-  ddsrt_nonnull((2))
-  ddsrt_attribute_format((printf, 2, 3));
-static int32_t idlc_parse(ddsts_type_t **typeptr);
+static int idlc_puts(const char *str, OUTDEST od);
+static int idlc_printf(OUTDEST od, const char *str, ...);
 
 #define CHUNK (4096)
 
 static int idlc_putn(const char *str, size_t len)
 {
-  assert(proc.state & IDL_WRITE);
+  assert(pstate->flags & IDL_WRITE);
 
   /* tokenize to free up space */
-  if ((proc.buffer.size - proc.buffer.used) <= len) {
-    if ((retcode = idl_parse(&proc)) == IDL_NEED_REFILL)
-      retcode = 0;
+  if ((pstate->buffer.size - pstate->buffer.used) <= len) {
+    if ((retcode = idl_parse(pstate)) == IDL_RETCODE_NEED_REFILL)
+      retcode = IDL_RETCODE_OK;
     /* move non-tokenized data to start of buffer */
-    proc.buffer.used =
-      (uintptr_t)proc.scanner.limit - (uintptr_t)proc.scanner.cursor;
-    memmove(proc.buffer.data, proc.scanner.cursor, proc.buffer.used);
-    proc.scanner.cursor = proc.buffer.data;
-    proc.scanner.limit = proc.scanner.cursor + proc.buffer.used;
+    pstate->buffer.used =
+      (uintptr_t)pstate->scanner.limit - (uintptr_t)pstate->scanner.cursor;
+    memmove(pstate->buffer.data, pstate->scanner.cursor, pstate->buffer.used);
+    pstate->scanner.cursor = pstate->buffer.data;
+    pstate->scanner.limit = pstate->scanner.cursor + pstate->buffer.used;
   }
 
-  if (retcode != 0)
+  if (retcode != IDL_RETCODE_OK)
     return -1;
 
   /* expand buffer if necessary */
-  if ((proc.buffer.size - proc.buffer.used) <= len) {
-    size_t size = proc.buffer.size + (((len / CHUNK) + 1) * CHUNK);
-    char *buf = ddsrt_realloc(proc.buffer.data, size + 2 /* '\0' + '\0' */);
+  if ((pstate->buffer.size - pstate->buffer.used) <= len) {
+    size_t size = pstate->buffer.size + (((len / CHUNK) + 1) * CHUNK);
+    char *buf = realloc(pstate->buffer.data, size + 2 /* '\0' + '\0' */);
     if (buf == NULL) {
-      retcode = IDL_MEMORY_EXHAUSTED;
+      retcode = IDL_RETCODE_NO_MEMORY;
       return -1;
     }
     /* update scanner location */
-    proc.scanner.cursor = buf + (proc.scanner.cursor - proc.buffer.data);
-    proc.scanner.limit = proc.scanner.cursor + proc.buffer.used;
+    pstate->scanner.cursor = buf + (pstate->scanner.cursor - pstate->buffer.data);
+    pstate->scanner.limit = pstate->scanner.cursor + pstate->buffer.used;
```

Comment:
I think this should be `pstate->scanner.limit = buf + pstate->buffer.used;`, but the `scanner.limit` is overwritten on line 100 anyway, so it doesn't really matter

Reply:
fixed